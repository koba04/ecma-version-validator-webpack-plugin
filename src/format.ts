import path from "path";
import { ErrorMap } from "./validate";

export const format = (errors: ErrorMap): string => {
  let message = "";
  errors.forEach((v, k) => {
    message += `${path.resolve(k)}\n`;
    let sourceIndents = 4;
    if (v.position) {
      const loc = `  ${v.position.line}:${v.position.column}`;
      message += loc;
      sourceIndents = loc.length + 2;
    }
    message += `  ${v.message}\n`;
    if (v.position) {
      message += `${" ".repeat(sourceIndents)}${extractSource(
        v.source,
        v.position
      )}\n`;
    }
  });
  return message;
};

const extractSource = (
  source: string,
  position: { line: number; column: number }
) => {
  const offset = 40;
  const sourceLine = source.split("\n")[position.line - 1];
  if (sourceLine.length < 100) {
    return sourceLine;
  }
  return sourceLine.substring(
    position.column - offset,
    position.column + offset
  );
};
