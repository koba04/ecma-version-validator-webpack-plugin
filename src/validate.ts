import { Source } from "webpack-sources";
import { parse, defaultOptions } from "acorn";
import { format } from "./format";

export type ECMAVersion = typeof defaultOptions.ecmaVersion;

export type ErrorMap = Map<
  string,
  {
    message: string;
    source: string;
    position: {
      line: number;
      column: number;
    } | null;
  }
>;

export const validate = (
  assets: { [file: string]: Source },
  options: { ecmaVersion: ECMAVersion; test: RegExp }
) => {
  const { ecmaVersion, test } = options;
  const errors: ErrorMap = new Map();
  Object.entries(assets)
    .filter(([file]) => test.test(file))
    .forEach(([file, source]) => {
      const sourceCode = source.source();
      try {
        parse(sourceCode, { ecmaVersion });
      } catch (e) {
        errors.set(file, {
          message: e.message,
          source: sourceCode,
          position: e.loc,
        });
      }
    });
  if (errors.size > 0) {
    throw new Error(format(errors));
  }
};
