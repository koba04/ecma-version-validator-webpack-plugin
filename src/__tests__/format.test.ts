import { format } from "../format";
import { ErrorMap } from "../validate";

jest.mock("path");

describe("format", () => {
  it("should return a formatted message", () => {
    const errorMap: ErrorMap = new Map();
    errorMap.set("foo.js", {
      message: "const is a reverved word",
      source: "var a = true;\nconst foo = () => {}",
      position: {
        line: 2,
        column: 0,
      },
    });
    expect(format(errorMap)).toMatchSnapshot();
  });
  it("should return a formatted message if position is null", () => {
    const errorMap: ErrorMap = new Map();
    errorMap.set("foo.js", {
      message: "const is a reverved word",
      source: "var a = true;\nconst foo = () => {}",
      position: null,
    });
    expect(format(errorMap)).toMatchSnapshot();
  });
  it("should omit a source code if the error line is over 100 characters", () => {
    const errorMap: ErrorMap = new Map();
    errorMap.set("foo.js", {
      message: "const is a reverved word",
      source: `var ${"a".repeat(100)} = () => {}; // ${"b".repeat(200)}`,
      position: {
        line: 1,
        column: 110,
      },
    });
    expect(format(errorMap)).toMatchSnapshot();
  });
});
