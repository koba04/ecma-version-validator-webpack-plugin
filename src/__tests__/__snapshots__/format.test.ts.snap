// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format should omit a source code if the error line is over 100 characters 1`] = `
"/home/dummy/foo.js
  1:110  const is a reverved word
         aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa = () => {}; // bbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
"
`;

exports[`format should return a formatted message 1`] = `
"/home/dummy/foo.js
  2:0  const is a reverved word
       const foo = () => {}
"
`;

exports[`format should return a formatted message if position is null 1`] = `
"/home/dummy/foo.js
  const is a reverved word
"
`;
